File: utils/src/main/java/uk/chaoticgoose/adventofcode/utils/grid2d/Point2D.java
package uk.chaoticgoose.adventofcode.utils.grid2d;

import org.jspecify.annotations.NullMarked;

import java.util.List;

import static uk.chaoticgoose.adventofcode.utils.grid2d.Direction2D.*;
import static uk.chaoticgoose.adventofcode.utils.grid2d.Direction2D.WEST;

@NullMarked
public record Point2D(int x, int y) {

    public Point2D toThe(Direction2D direction) {
        return switch (direction) {
            case North _ -> new Point2D(x, y - 1);
            case South _ -> new Point2D(x, y + 1);
            case East _ -> new Point2D(x + 1, y);
            case West _ -> new Point2D(x - 1, y);
        };
    }

    public List<Point2D> cardinalNeighbours() {
        return List.of(
            this.toThe(NORTH),
            this.toThe(EAST),
            this.toThe(SOUTH),
            this.toThe(WEST)
        );
    }
}

File: src/main/kotlin/utils/grids/twodee/Point2D.kt
package utils.grids.twodee

data class Point2D(val row: Int, val column: Int) {
    fun toThe(direction: Direction2D): Point2D {
        return when (direction) {
            Direction2D.North -> this.copy(row = row - 1)
            Direction2D.East -> this.copy(column = column + 1)
            Direction2D.South -> this.copy(row = row + 1)
            Direction2D.West -> this.copy(column = column - 1)
            Direction2D.NorthEast -> Point2D(row = row - 1, column = column + 1)
            Direction2D.SouthEast -> Point2D(row = row + 1, column = column + 1)
            Direction2D.SouthWest -> Point2D(row = row + 1, column = column - 1)
            Direction2D.NorthWest -> Point2D(row = row - 1, column = column - 1)
        }
    }

    fun neighbours(): List<Point2D> = listOf(
        this.toThe(Direction2D.North),
        this.toThe(Direction2D.NorthEast),
        this.toThe(Direction2D.East),
        this.toThe(Direction2D.SouthEast),
        this.toThe(Direction2D.South),
        this.toThe(Direction2D.SouthWest),
        this.toThe(Direction2D.West),
        this.toThe(Direction2D.NorthWest)
    )

    fun cardinalNeighbours(): List<Point2D> = listOf(
        this.toThe(Direction2D.North),
        this.toThe(Direction2D.East),
        this.toThe(Direction2D.South),
        this.toThe(Direction2D.West)
    )
}